File: 22_专题/区间问题/区间合并/enumerateConsecutiveSubarrays.go
package main

func main() {
	EnumetateConsecutiveSubarrays([]int{1, 2, 3, 5, 6, 7, 9}, 1, func(start, end int) {
		println(start, end)
	})
}

// 遍历连续的子数组
func EnumetateConsecutiveSubarrays(nums []int, diff int, f func(start, end int)) {
	i, n := 0, len(nums)
	for i < n {
		start := i
		for i < n-1 && nums[i]+diff == nums[i+1] {
			i += 1
		}
		i++
		f(start, i)
	}
}

File: 22_专题/区间问题/区间合并/enumerateConsecutiveSubarrays.py
from typing import Generator, List, Tuple


def enumetateConsecutiveSubarrays(
    nums: List[int], diff=1
) -> Generator[Tuple[int, int], None, None]:
    """遍历连续的子数组.

    >>> list(enumetateConsecutiveSubarrays([1, 2, 3, 5, 6, 7, 9]))
    [(0, 3), (3, 6), (6, 7)]
    """
    i, n = 0, len(nums)
    while i < n:
        start = i
        while i < n - 1 and nums[i] + diff == nums[i + 1]:
            i += 1
        i += 1
        yield start, i


if __name__ == "__main__":
    assert list(enumetateConsecutiveSubarrays([1, 2, 3, 5, 6, 7, 9])) == [(0, 3), (3, 6), (6, 7)]
